gomock/call.go

```diff
@@ -75,6 +75,9 @@ func (c *Call) Do(f interface{}) *Call {
 	return c
 }
 
+// Return sets the values to return when the target method is called.
+//
+// If Return and Do are called, the values returned via Do will be returned.
```

Comment:
@kaedys I agree with your analysis.

At this point, I am not sure if using this mock library warrants the extra complexity. Failures are cryptic and hard to resolve. I'm not sure if adding a new method that is subtly different from the others is a good solution either, as that will just result in similar frustration from others trying to add or modify unit tests.

This PR can either be closed or someone else may carry it.

The alternative that I'm using is described below.

---
At this point, I've just start writing mocks like this:

```go
type InterfaceUnderTest {
  MethodOne(...) (...)
  MethodTwo(...) (...)
}

type mockInterface struct {
  InterfaceUnderTest
}
```

Because the interface is embedded, the type implements the interface with no extra code. With a method-less definitions, it will panic over every method call, reported from the call site in the code under test. Tracking down the error during development is straightforward. When we want to add an implementation, we simply override the target method:

```go
func (m *mockInterface) MethodOne(...) (...) { return ... }
```

Now, when `(*mockInterface).MethodOne` is called, we get our implementation. When `(*mockInterface).MethodTwo` is called, we get a panic, which may be intended as part of the test, ensuring failure. If we need something more sophisticated, we can start adding additions to support what is needed for a particular set of tests.

Let's say we need to replaced the implementation of `(*mockInterface).MethodOne`:

```go
type mockInterface struct {
  InterfaceUnderTest

  MethodOneFn func(m *mockInterface, ...) (...) // signature must match, add receiver
}

func (m *mockInterface) MethodOne(m *mockInterface, ...) (...) { 
  return m.MethodOneFn(m, ...)
}
```

This let's us replace the implementation dynamically. This covers most of the cases I've needed but what about call ordering?

We can simply track the method call order and test against it later:

```go
type call struct {
  method interface{} // bound method instance
  args []interface{}    // arguments, as an interface slice
}

type mockInterface struct {
  InterfaceUnderTest
  MethodOneFn func(m *mockInterface, ...) (...) // signature must match, add receiver

  calls []call
}

func (m *mockInterface) MethodOne(m *mockInterface, ...) (...) { 
  m.calls = append(calls, call{m.MethodOne, []interface{...}})
  return m.MethodOneFn(m, ...)
}

func TestMyTest(t *testing.T) {
  // .. setup mocks
  if !reflect.DeepEqual(mock.calls, []call{
    {method: mock.MethodOne, args: []interface{"arg1", "arg2")}  
  }) {
    t.Fatalf("unexpected call chain")
  }
}
```

More sophisticated checking could be pulled up into a package. This would include helper functions to keep track of calls and match different styles of call chains, as well as an embeddable type for tracking and matching calls, similar to `testify/mock.Mock`. A lot of this can be codegen'd, as well but it is probably better to make the tests clear and explicit.